### ISSUE TITLE
CPDLC message type appears to be missing

### ISSUE THREAD

**Issue by robin24** [islandcontroller/hoppie-connector#1]
Hi,

First off, thanks for your work on this super handy package for Hoppie's ACARS!
While implementing this in a personal project, I kept wondering why I was unable to logon to a CPDLC station, e.g. KUSA.
I then found out that the request logon message needs to be sent as message_type: cpdlc, however it appears that this message type is missing from HoppieConnector.

I'm wondering if this is indeed correct, or whether I'm missing something here?

Thanks!

**Comment by islandcontroller** [comment]
Hi there!

Glad to see my stuff is useful to others :)
The message type is indeed not implemented! My primary use case was as an ACARS ground station, and I hadn't looked into CPDLC yet at that point.

Are you parsing the CPDLC message contents in your application? A generic CPDLC message container type should cover that usecase, I believe. 

Thanks for the feedback, I'll put this as a feature request!

**Comment by robin24** [comment]
You're more than welcome! :-)

Parsing and creating these CPDLC message payloads is the thing that I'm struggling with... I already tried creating a CPDLCMessage class based on how TelexMessage is implemented.
However, I then learned (the inner workings of CPDLC are still extremely new to me) that CPDLC messages must follow a specific format, which might look something like
```
/data2/6/19/N/WILCO
```
Unfortunately, I'm struggling to find out the meaning of those specifiers, all I know is that data2 refers to the payload format being used, and WILCO is the actual message being sent...

Do you by any chance happen to have an idea what the other stuff means?

Thanks!

**Comment by robin24** [comment]
OK... Was able to learn some more about the structure of these messages, but still not sure if my understanding is fully correct. Here's how it apparently works.
Given the following example message:
```
/data2/14//WU/CLD 1522 240826 EDDL PDC 005 OCN69FX CLRD TO @LICC@ OFF @23L@ VIA @DODEN9T@ CLIMB @500... [+59]
```
* The / character is used as the delimiter between the various data fields.
* data2 designates the data exchange format, this is always specified and always the same.
* 14 is the message ID. Can be randomly generated as far as I understand, doesn't have to be sequential or anything like that.
* // If this were a response to a previously received message, the ID of the referenced message would go here. No value in this case as the PDC message is the start of a message dialog.
* WU specifies the response choices given to the recipient, indicating how the message should be acknowledged. As far as I could figure out this can be one of:
* * NE: None Required
* * Y: Yes
* * N: No
* * WU: Wilco/Unable
* * AN: Affirm/Negative
* * R: Roger
* The final part of the payload is the actual message content.

Consequently, the response message to the above PDC looks like this:
```
/data2/5/14/N/WILCO
```

Just figured I'd leave my findings here, would be awesome if this could be implemented as a message type.

**Comment by islandcontroller** [comment]
I am not too familiar with the CPDLC format myself, but I did find a good source of documentation when writing this package: [devHazz/hoppie-acars-docs](https://github.com/devHazz/hoppie-acars-docs/blob/main/Messaging/CPDLC%20Format.md). Maybe this helps?
Other than that there's docs by Hoppie himself, but you need to use webarchive to get to them: [Hoppie's ACARS, CPDLC page](https://web.archive.org/web/20240419192409/https://www.hoppie.nl/acars/system/cpdlc.html) *(Archived April 2024)*.

As for the implementation, and going by the information provided in the links above, my idea would be to do some light pre-processing (i.e. splitting the message into min, mrn, ...) inside a `CpdlcMessage`. This would allow for some simple validation, too. Grabbing data from the plaintext message should probably end up as a RegEx in the application, anyway.

PS: I was still typing this as you posted your follow-up ;P Thanks for the info! It does line up with what I gathered so far. I will add the response choices to the pre-processing, probably an Enum type.

**Comment by islandcontroller** [comment]
Check the [`dev_cpdlc-message-type`](https://github.com/islandcontroller/hoppie-connector/tree/dev_cpdlc-message-type) branch whether this solution fits your usecase :)

Also, feel free to submit any "problematic" CPDLC messages as test cases when you encounter them in the wild!

Thanks for your help so far!

**Comment by robin24** [comment]
Wow that was lightning fast, you're awesome!!!
Haven't tested it yet but absolutely plan to do so tomorrow, but looking at the CpdlcMessage class this looks just perfect.
Also, really appreciate the docs you linked above, I didn't actually come across those pages on my own searches, the stuff that I did figure out came from me poking around in the public Hoppie's message log and correlating some data :-).

Huge thanks once again for implementing this, I'll run some tests ASAP and report back on how it went!

Cheers

**Comment by robin24** [comment]
Just had a chance to do some testing, and so far this is working perfectly!

```
>>> cnx.peek()
([(125, CpdlcMessage(from_name='KSFO', to_name='BVI4RK', min=9, rr=CpdlcResponseRequirement.WILCO_UNABLE, message='CLD TO @KJFK@. @TRUKN2@ DEPARTURE. MNTN @6000@. DPRTR ON @120.9@. SQUAWK @4514@. EXPCT RWY @01R@. ', mrn=None)), (128, CpdlcMessage(from_name='KUSA', to_name='BVI4RK', min=11, rr=CpdlcResponseRequirement.NOT_REQUIRED, message='LOGON ACCEPTED', mrn=1))]
```

Have yet to do a full flight with it and will be sure to keep looking out for weird edge cases as well, but so far it's absolutely fantastic! 👍

**Comment by islandcontroller** [comment]
Awesome! Thanks for the feedback! :)

I'll merge this in to master so it's available in the 0.2.0 release.
